File: models/Gpt-3.5-turbo/cpp/code/problems/additional_problems/Critical_Cities.cpp

#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> adj;
vector<bool> visited;

void dfs(int city) {
    visited[city] = true;
    for (int neighbor : adj[city]) {
        if (!visited[neighbor]) {
            dfs(neighbor);
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    adj.resize(n + 1);
    visited.resize(n + 1, false);

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
    }

    dfs(1);

    int count = 0;
    for (int i = 1; i <= n; i++) {
        if (!visited[i]) {
            count++;
        }
    }

    cout << count << endl;

    for (int i = 1; i <= n; i++) {
        if (!visited[i]) {
            cout << i << " ";
        }
    }

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/additional_problems/Critical_Cities.py

def dfs(city, adj_list, visited):
    visited[city] = True
    for neighbor in adj_list[city]:
        if not visited[neighbor]:
            dfs(neighbor, adj_list, visited)

n, m = map(int, input().split())
adj_list = [[] for _ in range(n+1)]

for _ in range(m):
    a, b = map(int, input().split())
    adj_list[a].append(b)

visited = [False] * (n+1)
critical_cities = set()

dfs(1, adj_list, visited)

for city in range(1, n+1):
    if not visited[city]:
        continue
    is_critical = True
    for neighbor in adj_list[city]:
        if not visited[neighbor]:
            is_critical = False
            break
    if is_critical:
        critical_cities.add(city)

critical_cities = sorted(critical_cities)

print(len(critical_cities))
print(*critical_cities)
